### DISCUSSION TITLE
Missing entities

### DISCUSSION THREAD

**Discussion by SteveEggo** [wills106/homeassistant-solax-modbus#558] [Q&A] [answered]
I have installed what I thought was homeassistant-solax-modbus it shows in HA Devices & Settings as Solax Power .
Have I actually got the correct integration ? 
I have configured it and have a hub set for serial and a sofar inverter
The inverter is connected via an isolated USB-RS485 converter. I can speak to the inverter via windows terminal program, so the hardware works)
I am running HA on a RaspberryPi, the converter is plugged into Pi and shows connected to '/dev/ttyUSB0'
I do not see any entities within the Solax Power Integration.
I want to read SOC, how do I set this up as an entity I can then display ? SOC available at reg address 0x9012
General question are there any examples I can download that show use of Solax Power.

**Comment by alienatedsec** [comment]
@SteveEggo which inverter?

**Reply by SteveEggo** [reply]
Sofar Solar HYD3600-ES Software version 1.0, Hardware version 1.0, with 2 x Pylontech US3000C batteries, 10 x 415W panels in 2 strings of 5

**Comment by SteveEggo** [comment]
More information now available. I have set up Solax to use the Sofar inverter type.
When HA startsup it sends a 8 byte message to the inverter.
01 03 04 45 00 07 14 ED (All shown in hex format)
This is a sensible message, its asking for the data from 7 registers starting at address 0x0445, its not getting a reply.
The documentation I have for the inverter does not have a register at 0x0445
Does anyone know what registers are supposed to be at this address.
Looking through what I can see of the Solax code I cannot locate 0x0445 so I need a bit more information.

**Reply by wills106** [reply]
Have you tried plugin_sofar_old ?
There are two different registers sets that I know off. Unless you have another unknown set?

If you have a working plugin it will return your Serial number in the error log even if the Integration doesn't fully work.

**Reply by SteveEggo** [reply]
I am having more luck with plugin_sofar_old. This reads the serial number from 0x2002.
There may be a bug in the inverter firmware that causes any read from 0x2000 to 0x200F to return the correct number of word but from address 0x2000 onwards. This has the effect of adding a 2 byte preamble to the serial number. So if the reply to read 0x2002 should be  SM1Exxxxxxxxx, it is actually \x00\x05SM1Exxxxxxxxx. The \x00\x05 is the product code from 0x2000.
I have got round this by adding:
elif 'SM1E' in seriesnumber:  invertertype = HYBRID | X1 # 3kW HYDxxxxES #SE Inverter replies with 2 byte prefex
to the end of the determineInverterType function within plugin_sofar_old
What I wanted to do was remove the 2 byte prefix with something like 
seriesnumber = seriesnumber[2:] but this causes an error  where everything became unavailable, error log does not show why.

I have set up a single card with all of the Sofar entities on it. I have a Sofar HYDxxxx-ES simulator that I have built in python. I am replying to the read of registers at 0x0200 with numbers that increment. This makes it easier to see what is and isn't working.
Out of 35 entities + 5 calculated I have a few failures as Entity not available: 
eps_voltage, eps_power, grid_voltage_s, grid_current_s, grid_voltage_t, grid_current_t
I have EPS disabled and the inverter is single phase so this is probably desired behaviour.
In the plugin_sofar_old file "PV Voltage 1" has key = "pv_oltage_1" note the missing v of voltage, I thought this was a typo but when I corrected it it broke the entity, it shows up as unknown ?
I would like to share my new version of plugin_sofar_old, shall I upload to your GITHub repo or would you prefer another way to share it. If you are interested in my simulator I can make that available.
I am using a simulator as I want to iron out all of the bugs in a controlled way until I get full control for Agile pricing and Inverter control. (That is still some way off.)

**Reply by wills106** [reply, marked as answer]
I think you have the same issue as https://github.com/wills106/homeassistant-solax-modbus/discussions/522#discussioncomment-7150932
I have just released [2023.10.1](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.10.1) which should return your Serial correctly.

**Reply by SteveEggo** [reply]
Tested 2023.10.1 and it does return a partial serial number, it is still prefixed with \x00\x05, but it does recognise it as a SM1E and assigns a type of HYBRID | X1
